### PR TITLE
Clean up parse_authenticate_body(), fix small bug. Add Unit Test. Improve OOB test.

### PR DESCRIPTION
**Pull request by sobomax**
This PR cleans up parse_authenticate_body() in the same way we've done with parse_qop_value(), fixes one error (accessing 1-st element of zero-lenth buf) and adds unit test.

I've also included some changes to improve OOB unit test output.

### CODE REVIEW

File: parser/test/test_parse_authenticate_body.c [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+		    "qop=auth,algorithm=MD5"), .tres = 0, .aflags = QOP_AUTH | AUTHENTICATE_MD5,
+		    .anonce = "esWk1wFa4bUBKzkmfKId++Y83eWzD9edBCGTwLV4Juk", .arealm = "[::1]"},
+		{.ts = str_init("Digest realm=\"sip.test.com\",qop=\"auth\",opaque=\"1234567890abcedef\","
+		    "nonce=\"145f5ca9aac6f0b9f93433188d446ae0d9f91a6ff80\",algorithm=MD5,stale=true"),
+		    .tres = 0, .aflags = QOP_AUTH | AUTHENTICATE_MD5 | AUTHENTICATE_STALE,
+		    .anonce = "145f5ca9aac6f0b9f93433188d446ae0d9f91a6ff80", .aopaque = "1234567890abcedef",
+		    .arealm = "sip.test.com"},
+		{.ts = str_init("DiGeSt\r\n\trealm=\"a\",\r\n\tqop=\"auth-int, auth\",\r\n\tnonce=\"n\",\r\n\topaque=\"0\",\r\n\t"
+		    "algoriTHm=md5"), .tres = 0, .aflags = QOP_AUTH | AUTHENTICATE_MD5 | QOP_AUTH_INT,
+		    .anonce = "n", .aopaque = "0", .arealm = "a"},
```
Comment (liviuchircu) [inline parser/test/test_parse_authenticate_body.c:59]: All is good here -- but here is some personal feedback (nitpicking):

* I found it a bit hard to tell why each `-1` test was expected to fail.  Eventually, I understood that it was due to the new RFC 8780 algorithms
* the entire test set looks like a wall of text and is hard to read.  Maybe adding an empty-line and/or comments between each test would mitigate this problem?

File: parser/test/test_oob.h [hunk 2]
```diff
@@ -18,4 +18,6 @@
  * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301,USA
  */
 
-void test_oob(const str *, void (*)(const str *, void *), void *);
+enum oob_position {OOB_PRE, OOB_POST};
```
Comment (liviuchircu) [inline parser/test/test_oob.h:21]: I'm not a big fan of this enum, as its purpose is not immediately obvious.  Would you also agree that these alternatives would be net superior, or is it just me?

* `enum oob_test {OOB_UNDERFLOW, OOB_OVERFLOW}`
* `enum oob_test {OOB_UNDERRUN, OOB_OVERRUN}`

File: parser/test/test_parse_qop.c [hunk 3]
```diff
@@ -65,10 +65,12 @@ void test_parse_qop_val(void)
...
-void test_parse_qop_oob(const str *tstr, void *farg)
+void test_parse_qop_oob(const str *tstr, enum oob_position where, void *farg)
 {
 	struct authenticate_body *ap = farg;
 
 	parse_qop_value(*tstr, ap);
-	ok(1, "oob check: parse_qop_value(\"%.*s\")", tstr->len, tstr->s);
+	ok(1, "oob check: parse_qop_value(%s\"%.*s\"%s)",
+	    where == OOB_PRE ? "->" : "", tstr->len, tstr->s,
+	    where == OOB_POST ? "<-" : "");
```
Comment (liviuchircu) [inline parser/test/test_parse_qop.c:75]: Code smell:

These 3 lines are copy-paste from `test_parse_authenticate_body_oob()`.  Depending on how this code is going to evolve, maybe this `"->"` and `"<-"` printing code will give more value/reusage if placed in `test_oob()`, with some parameter refactoring, if needed.

**Review by liviuchircu — CHANGES_REQUESTED** [review]

**Comment by sobomax** [comment]
Liviu, thanks for a review. OK, all 5 suggestion have been incorporated. Please check.

**Review by liviuchircu — APPROVED** [review]
Thank you, @sobomax!

**Pull request merged**